Zaibatsu

romainl opened the pull request:
Zaibatsu is liberally inspired by the so-called "[neo-tokyo](https://www.google.com/search?q=neo+tokyo+aesthetic&tbm=isch&ved=2ahUKEwjU89Cm2f78AhWmnycCHTyNAkgQ2-cCegQIABAA&oq=neo+tokyo+aesthetic&gs_lcp=CgNpbWcQAzIFCAAQgAQyBQgAEIAEMgUIABCABDIGCAAQBRAeMgYIABAIEB46BggAEAcQHjoICAAQBRAHEB46CAgAEAgQBxAeUKcKWIgjYKImaABwAHgAgAE5iAG0BJIBAjExmAEAoAEBqgELZ3dzLXdpei1pbWfAAQE&sclient=img&ei=XcvfY5TRAaa_nsEPvJqKwAQ&bih=1013&biw=1800&rlz=1C5GCEM_en)", "[neon-noir](https://www.google.com/search?q=neon+noir+aesthetic&tbm=isch&ved=2ahUKEwjc88Sk2f78AhWob_EDHbCPAnkQ2-cCegQIABAA&oq=neon+noir+aesthetic&gs_lcp=CgNpbWcQAzIFCAAQgAQyBggAEAgQHjIGCAAQCBAeMgYIABAIEB4yBggAEAgQHjIGCAAQCBAeUJoGWJoGYN0KaABwAHgAgAEyiAFZkgEBMpgBAKABAaoBC2d3cy13aXotaW1nwAEB&sclient=img&ei=WMvfY9znJqjfxc8PsJ-KyAc&bih=1013&biw=1800&rlz=1C5GCEM_en)", "[cyberpunk](https://www.google.com/search?q=cyberpunk+aesthetic&tbm=isch&ved=2ahUKEwihuOWw2f78AhVgW_EDHdgDDYcQ2-cCegQIABAA&oq=cyberpunk+aesthetic&gs_lcp=CgNpbWcQAzIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQ6BAgAEEM6BggAEAcQHjoGCAAQCBAeUNsKWOoZYIweaABwAHgAgAE3iAHwA5IBAjEwmAEAoAEBqgELZ3dzLXdpei1pbWfAAQE&sclient=img&ei=csvfY6G3FOC2xc8P2Ie0uAg&bih=1013&biw=1800&rlz=1C5GCEM_en)", "[vaporware](https://www.google.com/search?q=vaporware+aesthetic&tbm=isch&ved=2ahUKEwjNq-e72f78AhV_nycCHXT3BSQQ2-cCegQIABAA&oq=vaporware+aesthetic&gs_lcp=CgNpbWcQAzIFCAAQgAQyCQgAEIAEEAoQGDoECAAQQzoGCAAQBxAeUPkIWIsfYNQkaABwAHgAgAFAiAGOBJIBAjEwmAEAoAEBqgELZ3dzLXdpei1pbWfAAQE&sclient=img&ei=icvfY43DGv--nsEP9O6XoAI&bih=1013&biw=1800&rlz=1C5GCEM_en)", etc. a e s t h e t i c, as well as popular colorschemes like [dracula](https://github.com/dracula/vim), [tokyonight](https://github.com/folke/tokyonight.nvim), or [onedark](https://github.com/joshdick/onedark.vim).

[Here is the palette](https://color.hailpixel.com/#510039,00AF5F,FFAF00,5F00D7,D700FF,5FD7D7,D75F87,5FFF87,FFFF5F,AFAFFF,FFAFFF,5FFFFF,FFFFFF,DDDBDF,BAB6BF,98929F,766D7F,54495F,31243F,0F001F) at the time of this writing: 
<img width="1610" alt="Capture d’écran 2023-02-05 à 16 33 37" src="https://user-images.githubusercontent.com/344335/216828829-29b1896f-e0cd-4212-9d48-c6cd1d3d413d.png">

Of note: the colors don't change much between GUI and 256c except for the grayscale ramp which gets a slight purple-ish hue in GUI. Because the darkest of these is used as `Normal` background, the GUI experience is more nuanced than in 256c: very dark purple background versus black background.

This is mostly to get feedback. It has become quite usable in GUI and 256c but there is a lot of work left for 16c and 8c.

The screenshots below are all taken in MacVim on macOS:
<img width="1027" alt="Capture d’écran 2023-02-05 à 16 11 41" src="https://user-images.githubusercontent.com/344335/216827913-60a0c564-3136-4b59-87ae-b369d81572a5.png">
<img width="1012" alt="Capture d’écran 2023-02-05 à 16 12 25" src="https://user-images.githubusercontent.com/344335/216827916-7a1e0037-d5c3-4e35-bc19-0db3840e9167.png">
<img width="1027" alt="Capture d’écran 2023-02-05 à 16 13 04" src="https://user-images.githubusercontent.com/344335/216827921-0ac70cb4-fb05-4fc1-89fe-d6a056e625a6.png">
<img width="1027" alt="Capture d’écran 2023-02-05 à 16 15 38" src="https://user-images.githubusercontent.com/344335/216827927-3ce11cd8-c00b-4c66-abae-62620d2b5cd7.png">
<img width="1012" alt="Capture d’écran 2023-02-05 à 16 44 24" src="https://user-images.githubusercontent.com/344335/216829397-d2b6c3dc-870d-47e1-9824-980b37a9c955.png">

habamax commented:
It looks complete to me in 16/256/gui.
The vibe is there:
![image](https://user-images.githubusercontent.com/234774/216848776-5ec4c573-d3c4-4d16-9ae6-52199ffe850c.png)

8c is not there yet I guess? Or is it just a BG issue?
![image](https://user-images.githubusercontent.com/234774/216848819-9e399c2f-7aa5-49e3-8cdf-ec379d947c06.png)

romainl commented:
16c is more or less functional but it needs tweaking. I didn't work on 8c so far.

romainl commented:
OK, from lo-def to hi-def…

8c:
<img width="885" alt="Capture d’écran 2023-02-06 à 09 05 45" src="https://user-images.githubusercontent.com/344335/216917697-7f272e3c-bd04-4667-9c82-bb4a79a2c9d2.png">

16c:
<img width="885" alt="Capture d’écran 2023-02-06 à 09 07 32" src="https://user-images.githubusercontent.com/344335/216917943-3725f518-bc2f-4fa4-9851-d1243a779098.png">

256c:
<img width="885" alt="Capture d’écran 2023-02-06 à 09 08 50" src="https://user-images.githubusercontent.com/344335/216918187-7b3037b3-e849-413f-b2dc-586ad5c2c265.png">

GUI:
<img width="885" alt="Capture d’écran 2023-02-06 à 09 10 00" src="https://user-images.githubusercontent.com/344335/216918383-a08b4a32-5ab0-42bf-a11b-d84914d7f608.png">

habamax commented:
8c statuslinenc, is it intended?

![image](https://user-images.githubusercontent.com/234774/216943818-fe1a522c-2acf-4901-a24f-5953025c510a.png)

romainl commented:
Intended? Yes. Satisfactory? No.

habamax commented:
currently it looks not very good for a hor split:
![image](https://user-images.githubusercontent.com/234774/216955206-0dd101ee-a213-43a0-86bf-c551adff07fd.png)

What about this?
![image](https://user-images.githubusercontent.com/234774/216955804-92e5d7a8-721e-4004-b948-e4acceab1d9e.png)

```
hi Statusline ctermfg=grey ctermbg=black cterm=bold,reverse
hi StatuslineNC ctermfg=black ctermbg=grey cterm=NONE
```

neutaaaaan commented:
Documenting feedback I gave to @romainl in #vim :

The only 2 problems I have with `zaibatsu` after a cursory glance have to do with contrast: `Type` and `Preproc` are noticeably darker than the rest of the palette, and `Preproc` clashes pretty  badly with `Identifier` in some contexts.

`Type` could be brought up from `ctermfg=168` to `ctermfg=205` (or I suppose any other shade of red with a similar perceived brightness).
![image](https://user-images.githubusercontent.com/35198994/216980818-1b84d3d6-2b66-48c3-80d0-37cd911ffac9.png)
![image](https://user-images.githubusercontent.com/35198994/216980884-1824edd2-f8e5-49eb-b1fa-4787455ca0ba.png)

`Preproc` at `ctermfg=80` simply doesn't survive my usual torture test. It's very contrived, but I've found it to be a good  indicator of whether `Preproc` would also survive being used very sparingly, as I've been trying to work around of a similar clash between `Preproc` and `Statement` in [Iosvkem](http://github.com/neutaaaaan/iosvkem) for several years now.
![image](https://user-images.githubusercontent.com/35198994/216984236-b9639e82-753b-42fa-a716-227a94c386df.png)
![image](https://user-images.githubusercontent.com/35198994/216984296-d1583182-824d-4c41-8217-3307b891b5b0.png)

I should be able to take a closer look at it in the evening, the room I am currently sitting in is very bright.

romainl commented:
@neutaaaaan I have made the following tweaks:

- Type is brighter
- Identifier is a bit brighter
- PreProc is more blue, making it easier to differentiate from Identifier
- Special is more green, making it easier to differeniate from Identifier
- The neutralX colors in 256c now have a purple-ish hue too. It is less subtle than in GUI but it seems to work

Left: GUI, right: 256c
<img width="1799" alt="Capture d’écran 2023-02-07 à 08 16 54" src="https://user-images.githubusercontent.com/344335/217176522-ea6a0d07-f133-4e55-b756-bfa2ed0b6b68.png">

neutaaaaan commented:
`Identifier` and `Special` are still hardcoded to 87 and 84 in your template and the resulting colorscheme, not that I can really tell the difference anyway.

I'm not crazy about the shade of blue used for `PreProc`, but I'm not crazy about *any* of the shades of blue that I tried either, and it's one of the more balanced ones, so I suppose it'll have to do. I assume the colorschemes you referenced have fairly similar issues ?  
I'm staring at [this](https://user-images.githubusercontent.com/35198994/217203289-81fef8f8-7d2e-42b3-86e6-3cafc528f214.png) right now but this specific shade of *low-grade depression* doesn't exactly remind me of a mighty dub, more so of being slightly hungover on a sunday morning, waiting for *The Deliverator* to show up, *a very specific Manu Chao song* being played on a loop somewhere in the neighbourhood.

Anyway, that's enough bikeshedding the yak's haircut for now.

habamax commented:
Not sure if cursorline is too distinct in 256 compared to gui:

256

![image](https://user-images.githubusercontent.com/234774/217375997-74f2ae84-35a1-4485-9c47-e8ef61231c7f.png)

gui
![image](https://user-images.githubusercontent.com/234774/217376114-46e08a93-9ecf-4913-bf51-758f6fcf485b.png)

habamax commented:
Also StatuslineNC could be "more the same" between GUI and 256. In GUI it is grey, in 256 has some bluish tint.

habamax commented:
Another thing to mention is DiffText + Cursorline:
![image](https://user-images.githubusercontent.com/234774/217664102-797d06f6-ef88-4894-8e25-9287220027c7.png)

craigmac commented:
I love it. I'm made something similar for myself but nowhere near the quality of detail here, I'll be using it daily so feedback incoming as I use it. 

Couple things:, I like doing `hi! link CurSearch IncSearch`, what are you feelings on that? Second, the g:terminal_ansi_colors seem too dark to use in Alacritty (with and without tmux). Have a look at the PS1 colors and if I run something like lazygit or tig the red is too dark, it seems like the colors 8-15 would work better as the 0-7 colors but that's not how it's done usually, the 0-7 colors being the bright versions.

<img width="1156" alt="Screen Shot 2023-02-08 at 9 31 22 PM" src="https://user-images.githubusercontent.com/7416158/217702480-75e9817a-e1a7-4e86-b25f-b0ada54a78da.png">

romainl commented:
@craigmac Thanks. Feel free to report any issue you might have.

> I like doing `hi! link CurSearch IncSearch`, what are you feelings on that?

A link to `IncSearch` makes a lot more sense to me. `CurSearch` being linked to `Search` by default is a bit weird and Bram's performance concerns don't seem justified to me but we are kind of following `default`'s pattern everywhere in the name of consistency.

> the g:terminal_ansi_colors seem too dark to use in Alacritty (with and without tmux).

IIRC I allocated the ANSI colors right at the beginning but yeah, the hexadecimal value changed a bit in the meantime so the assignation is not good anymore. The dark red being very obvious.

I will fix that, thanks.

romainl commented:
@habamax initially, `StatusLineNC`, `ToolbarButton`, and `CursorLine` were neutral gray in 256c and closer in brightness to their GUI counterpart. The current version tries to be closer in terms of hue (purple-ish) but it is at the cost of brightness similarity because of how the xterm color cube works. On one hand it brings the whole 256c look and feel closer to GUI but on the other hand the colors are clearly too bright.

I may keep the purple-ish hue for `StatusLineNC` and `ToolbarButton` but `CursorLine` is way too bright so it will probably go back to a dark neutral gray.

neutaaaaan commented:
![image](https://user-images.githubusercontent.com/35198994/217767142-49c9efcf-4b11-4645-a589-cdf32c634413.png)

I rigged this up using #d70087 and #5f5ff, seems like they could work.
@romainl Is the xterm vibe, with the screaming bright green, yellow and cyan next to that pale light blue on purpose ? I don't really trust the built-in terminal with any of this.

habamax commented:
> I may keep the purple-ish hue for `StatusLineNC` and `ToolbarButton` but `CursorLine` is way too bright so it will probably go back to a dark neutral gray.

That is okay to have purple-ish hue for those, I pointed out that you have it in 256c but not in GUI...

habamax commented:
![image](https://user-images.githubusercontent.com/234774/217995966-812c4fc4-b37c-4867-a74c-60e8e1849f43.png)

![image](https://user-images.githubusercontent.com/234774/217996024-ead821b6-e682-4d31-9f7d-1b301c3d940a.png)

romainl commented:
No, it's not just gray, it has a slight purple hue: `neutral3` is #98929f.

habamax commented:
> No, it's not just gray, it has a slight purple hue: `neutral3` is #98929f.

Oh, my bad, didn't check the actual value -- it just looked so regularly gray so I blindly assumed it is gray.

romainl commented:
As I was working on fixing the term colors, I noticed that my prompt in the MacVim GUI didn't have the expected colors. Would you have an idea of why? `$TERM` is `xterm-256color`.

Expected:

<img width="360" alt="Capture d’écran 2023-02-10 à 09 09 05" src="https://user-images.githubusercontent.com/344335/218037239-dcf88f1d-6542-43c3-8b19-7c9a85d59073.png">

Actual:

<img width="362" alt="Capture d’écran 2023-02-10 à 09 08 55" src="https://user-images.githubusercontent.com/344335/218037344-27342329-0439-4ca3-84df-9f0f4095c29e.png">

The ability to display colors s definitely there, though:

<img width="569" alt="Capture d’écran 2023-02-10 à 09 12 09" src="https://user-images.githubusercontent.com/344335/218037563-89a657f8-9f0d-453c-a8e7-075f0e06c1b7.png">

craigmac commented:
Small nits (while using termguicolors):

Errormsg:  readable but the contrast isn't great: 
<img width="799" alt="Screen Shot 2023-02-10 at 11 49 39 AM" src="https://user-images.githubusercontent.com/7416158/218148949-5788120b-cfe3-4b81-99f0-a7315e9e5777.png">

SpellBad and friends: can we use `cterm=undercurl` and or is that too risky? Docs say it will fallback to underline if it isn't supported, but I'm thinking of situations where maybe user is saying it's supported but did not provide the correct `t_Cs` and `t_Ce` values as suggested in `:h underline-codes`. I haven't tested whether Vim looks just assumes those are correct and you end up with no undercurls OR underlines, or what it does. I have the same questions about `highlight-ctermul` and if

neutaaaaan commented:
@craigmac Interactions between vim and terminal emulators are just too finnicky to risk using undercurls. For example, the fallback doesn't work at all using tmux 3.3a running inside urxvt 9.31, but it works inside urxvt itself.

romainl commented:
> `ErrorMsg`

I switched `Error` to the `error` color (pure red) but forgot `ErrorMsg`. Will fix this ASAP.

> `undercurl`

Indeed it doesn't work in lots of places and the fallback to `underline` isn't great either.

> `highlight-ctermul`

Same story.

romainl commented:
So I have made a bunch of tweaks:

- ANSI 1 and ANSI 4 are more legible in `:term`
- 256c: `CursorLine` is a flat dark gray
- 256c: This makes it possible to darken the other blue-ish "neutral" hues
- add proper `StatusLineTerm/NC`
- fix `ErrorMsg`
- fix ColorColumn bleedind through Folded

romainl commented:
256c:
<img width="883" alt="Capture d’écran 2023-02-10 à 22 07 55" src="https://user-images.githubusercontent.com/344335/218197854-15e64282-a431-4699-851a-daf5972177ad.png">

- the cyans are less jarring
- the greens are warmer, which helps differrentiatiing them with cyans

romainl commented:
Do you think the background is purple enough in GUI?

habamax commented:
@romainl it is enough purple for me

habamax commented:
Another thing to check is QuickFixLine vs IncSearch:
https://asciinema.org/a/XNu0mcn0kaaQE46XwQCmeN2BY

craigmac commented:
I liked the old green, this one reads more sickly than neon to me, maybe because I'm just fond of that tmux green 😀 And was going to say about the background, I think I'd like it a little more purple, somewhere between the original background and this latest one

romainl commented:
So…

- I shifted the GUI background a little toward blue, as well as the "neutrals" that are derived from it.
- I tried a bunch of slightly brighter values for the GUI background but it was always too bright for my taste. I think I will stop playing with the GUI background, now. If you find the time to experiment with alternative values (remember that `neutral1-5` are derived from `black`), feel free to post screenshots.
- I handled the `QuickFixLine` vs. `IncSearch` business by making `IncSearch` "black" on white. After 30 minutes I couldn't find a better `QuickFixLine` and everything I tried broke in one way or another in 16c. The new `IncSearch` doesn't clash with `QuickFixLine` anymore and it looks basically the same across environments.
- I think that green will stay. Insisting on keeping it slightly blue-ish only causes race conditions with cyans and I don't want to go there again.

GUI:
<img width="1012" alt="Capture d’écran 2023-02-11 à 09 54 14" src="https://user-images.githubusercontent.com/344335/218249548-f49e4497-0ebf-4432-ba4f-66dbdff062c5.png">

256c:
<img width="881" alt="Capture d’écran 2023-02-11 à 09 55 37" src="https://user-images.githubusercontent.com/344335/218249570-f16f4988-f96e-47b1-8298-ea0b8d7789fb.png">


FWIW, I tried zaibatsu in 16c with my decades-old custom terminal palette and I think it works pretty well (even if @neutaaaaan will find it too muted):
<img width="885" alt="Capture d’écran 2023-02-11 à 09 31 03" src="https://user-images.githubusercontent.com/344335/218249193-715b0455-5bc8-4a56-a924-5d6fc03dcacf.png">

Even though I use the default xterm palette as a baseline when working on 16c, it can be useful to check with other palettes. That said, I dread the test with Solarized ;-)

neutaaaaan commented:
This really goes to show color perception depends on the particular combination of eyeballs and display.

`Special` looks sickly green, bordering on yellow to me, and collides pretty badly with`Constant` 
![image](https://user-images.githubusercontent.com/35198994/218252638-caa4cea8-01fa-491d-a36a-256119fdd124.png)
![image](https://user-images.githubusercontent.com/35198994/218252401-8e02da64-b501-430c-ba45-7696222d7032.png)

Using `#87ff00` instead of `#afff00` seems more balanced to me, without sacrificing the perceived brightness, or colliding with `Preproc`.
![image](https://user-images.githubusercontent.com/35198994/218252659-29980470-928d-4f4a-93dd-7623e3303847.png)
![image](https://user-images.githubusercontent.com/35198994/218252459-6c3e26ec-d7ff-4a02-a2ba-a9f597a1310c.png)

2 issues with the built-in term in gvim :
![image](https://user-images.githubusercontent.com/35198994/218252774-4648df13-c5bd-4280-82f3-3b08c01492d4.png)
That dark red just doesn't work, and the `StatusLine` is going to be mighty confusing for `tmux` users.

romainl commented:
- Hmm… OK for #87ff00.
- I will see what I can do for the dark red in `:term`.
- I went with the default logic for `StatusLineTerm`, having no idea how to denote a `:term` or even if that makes sense. Should I simply link them to their non-term equivalent or is there a better strategy?

Indeed, color is complicated.

neutaaaaan commented:
I had written a much longer post about the terminal colors, but I think most of the issues I have with them stem from the reference palette being the brighter set of colors. I figure I might as well just try and do it "my way" sometimes tomorrow and see if any useful feedback comes out of that.

> The new IncSearch doesn't clash with QuickFixLine anymore and it looks basically the same across environments.

The new `IncSearch` blends in way too much to really be useful if you're not actively looking for it.
![image](https://user-images.githubusercontent.com/35198994/218274083-eb69785a-d4c1-4206-a63e-22812cc9232d.png)
![image](https://user-images.githubusercontent.com/35198994/218274144-43467c2f-7fc0-4edd-8d09-d7fddd41ac16.png)

habamax commented:
Am I missing the point? I have a different incsearch which standouts quite nicely:
![image](https://user-images.githubusercontent.com/234774/218289114-ee8de662-8122-4c6c-8097-408f066f7511.png)

Both in gui and 256c...

habamax commented:
oh, I was one iteration behind...

White is ok to me.

neutaaaaan commented:
After messing about with the terminal palette for a while, and coming up with something that I feel works a bit better, I believe the biggest issue with `zaibatsu` is having to work from a colorscheme that uses very bright colors *down* to the dark palette that the built-in terminal will always stick to in my experience.  
Making the bottom half more balanced means fiddling with the top half too so both sets of colors don't feel completely divorced, and not only is the end result not that great because of the limitations inherent to the xterm palette, it probably doesn't really look like like `zaibatsu` as it is now.

Would such a disconnect be acceptable ?

All of my experimenting made me realize that `:term` handles bold/bright by using a bold typeface, which can't be assumed to even exist, whereas the bright colors *can* be relied on. What this means, is that the top half of the palette is never seen, which in the worst case scenario means information loss, but in the case of `zaibatsu` means `:term` effectively hides most of the disconnect I mentionned before. Has this been brought up before ?

romainl commented:
Of course having a `:term` palette that closely matches the colorscheme's palette would be perfect but I am definitely open to having _some_ differences if that means more usability.

neutaaaaan commented:
Here's the results after a few passes through the grinder.
![image](https://user-images.githubusercontent.com/35198994/218312543-76537c49-755b-4b0c-8fff-19b5ae381db9.png)

This seems too much like `quiet` too me. Maybe working from an only *slightly* tamed bright palette down to a *slightly* brightened dark palette would turn out better.

habamax commented:
@romainl what about this difftext?

![image](https://user-images.githubusercontent.com/234774/218587992-74491a7e-ed98-41f1-9411-85c19a8ee0e0.png)

romainl commented:
FWIW, [this](https://color.hailpixel.com/#510039,5FAF00,FFAF00,5F00D7,D700FF,5FD7FF,FF5FAF,87FF00,FFFF5F,AFAFFF,FFAFFF,87FFFF,FFFFFF,D7D5DB,AFAAB6,878092,5E556D,362B49,0E0024,D70000) is the current "main" palette:
<img width="1619" alt="Capture d’écran 2023-02-14 à 21 00 19" src="https://user-images.githubusercontent.com/344335/218848385-24fbd06a-20c5-4043-8b5a-99f563dda1c1.png">

@moshe, could you post the values you used for the latest screenshot? I would like to see if we can meet halfway :-).

@habamax, it might work in GUI because of the definition but I doubt it is going to work in 16c, for example.

To be honest, I am not super happy with that diff and most importantly how `CursorLine` interacts with `DiffAdd` and friends. It works incredibly well in 16c but not at all in 256c and GUI. Bummer.

habamax commented:
@romainl that is actually quite ok:
![image](https://user-images.githubusercontent.com/234774/218865853-92ebdc51-b3c9-49ef-b4f3-ff65b724232d.png)

romainl commented:
@habamax I don't like it.

Let's backtrack a little…

- Without `cterm=reverse`, `DiffAdd` and friends are eaten away by `CursorLine`, which makes it hard to figure out if the cursor is in a diff and so on.
- With `cterm=reverse`, the background color of `DiffAdd` and friends stays so we always have the diff context, which is the most important info in a diff, but the "reverse" bit is also applied to `CursorLine` so the actual effect is: the correct bg from `DiffX` + the bg of `CursorLine` for fg.
- It works if the fg of `DiffX` is already either the same as `CursorLine` bg or something closer.
- It doesn't work if they are too different because the fg change might make it less legible.
- In my own accessible diff colorscheme, the fg of `DiffX` is dark so the effect of `CursorLine` is either invisible or negligible.
- In the default diff colorscheme, the fg of `DiffX` is light so the effect of `CursorLine` is too pronounced and I find it jarring.

FWIW, this made me spot a couple of issues in my own colorschemes. BRB.

![cl](https://user-images.githubusercontent.com/344335/218947031-dacbfda8-5b09-43e0-ba59-c206f41190a9.gif)

habamax commented:
It is not very appealing, but at least it became readable.

For wildcharm I use:
![image](https://user-images.githubusercontent.com/234774/218956397-5dee6454-c20c-46c9-9f3e-d483febab748.png)

And it let me see the Difftext in CursorLine.

neutaaaaan commented:
@romainl Highlighting `moshe` instead of @neutaaaaan is one hell of a fingerslip.

```
black: #0e0024 
darkred: #d7005f 
darkgreen: #00af5f
darkyellow: #d78700
darkblue: #5f5fff
darkmagenta: #d700ff
darkcyan: #5fd7ff
darkwhite: #d7d5db
grey: #878092
brightred: #ff0087
brightgreen: #00d700
brightyellow: #ffd700
brightblue: #8787ff
brightmagenta: #ff87ff
brightcyan: #00ffff
brightwhite: #ffffff
```
I reckon darkgreen, darkcyan, and maybe even darkyellow can be salvaged, but I probably pushed down too hard on the bright colors.

romainl commented:
In the last commit…

- background is transparent for 8c 16c and 256c
- StatusLineTerm/NC is now a link to StatusLine/NC
- VisualNOS is more readable
- small tweaks

The biggest change is the background. My idea is that zaibatsu is meant to have a dark purple-ish background,  something that can only be achieved in a declarative way in the GUI or in a termguicolors-compatible terminal. In other cases (8c, 16c, 256c w/o termguicolors) I wonder if it is justifiable to make it transparent and let crafty users adjust their terminal's background as they see fit.

The screenshots below are taken in a 256c terminal emulator. The first one uses my default terminal's background, a dark gray, the second one uses pure black, the third one uses the same color as the GUI version, the last one is a bit on the insane side of things.

<img width="1152" alt="Capture d’écran 2023-02-15 à 10 29 13" src="https://user-images.githubusercontent.com/344335/218991848-bf17d25f-b063-40fb-b761-8176c0e07718.png">

<img width="1152" alt="Capture d’écran 2023-02-15 à 10 29 40" src="https://user-images.githubusercontent.com/344335/218991904-844fcbff-0880-403f-b442-75d085522cc8.png">

<img width="1152" alt="Capture d’écran 2023-02-15 à 10 30 33" src="https://user-images.githubusercontent.com/344335/218991948-a096b3ae-0ab0-484d-ac35-449175c881f2.png">

<img width="1152" alt="Capture d’écran 2023-02-15 à 10 30 21" src="https://user-images.githubusercontent.com/344335/218992003-9dfba1b6-5c03-4114-afca-3ad75e8cc38b.png">


What do you think?

habamax commented:
> What do you think?

We have no colorschemes with transparent bg in the repo. Imagine transparent peachpuff on a black terminal or transparent zaibatsu on a gray/white bg terminal.

neutaaaaan commented:
I'm not crazy about the idea either, it's bound to generate false positives.

romainl commented:
OK, reverted.

habamax commented:
Spent a day with zaibatsu, looks good to me.

habamax commented:
Hm, today I am using a different monitor and it has probably different settings... the blue vs cyan is a bit annoying, being too similar:

![image](https://user-images.githubusercontent.com/234774/219519126-f99e6f25-14a9-4074-8c57-ea4c53da6d36.png)

Like it is almost the same:
![image](https://user-images.githubusercontent.com/234774/219519568-d3ef32b8-adc3-40ac-ab9d-2aa0a7237d6b.png)

Again, maybe this is monitor settings issue and on yours it would do better.

romainl commented:
Monitors are different, monitor settings are different even for a same monitor, lighting conditions are different, eyeballs are different, fonts are different, font smoothing technologies are different, brains are different, etc. We can't possibly hope to have an optimal outcome in every possible scenario and we need to accept that.

Anyway, my options are limited…

#00AFFF is too "electric blue" IMO but well contrasted:
<img width="369" alt="Capture d’écran 2023-02-17 à 08 06 53" src="https://user-images.githubusercontent.com/344335/219573900-60a3ed10-a010-472d-a226-4fbfed153201.png">

#00D7ff:
<img width="452" alt="Capture d’écran 2023-02-17 à 08 08 51" src="https://user-images.githubusercontent.com/344335/219574156-daea21ab-e410-4ede-885b-54283cb2d0ef.png">

#00AFD7 is well contrasted but too close to "dark cyan":
<img width="417" alt="Capture d’écran 2023-02-17 à 08 10 46" src="https://user-images.githubusercontent.com/344335/219574754-7a1380ba-60c3-4e82-b330-2a0e19275c87.png">

#5FD7FF is the current value:
<img width="506" alt="Capture d’écran 2023-02-17 à 08 12 18" src="https://user-images.githubusercontent.com/344335/219574905-31ac3f23-5b73-47b5-9b46-96045eb75934.png">

habamax commented:
first and third looks way better for me...

neutaaaaan commented:
I'm fine with what we have now: there's no way around this collision that doesn't involve [stepping out of the palette entirely](https://user-images.githubusercontent.com/35198994/219660694-c0a0b554-7f9f-4929-a41c-0533e8c460ff.png).

The white `IncSearch` I'm still not crazy about.

romainl commented:
In the last couple of commits, I…

- changed darkcyan to #00AFFF,
- adjusted `Visual`,
- changed `IncSearch` to black on orange/darkyellow (similar to `Folded` but they can't clash as there is no search highlighting in closed folds),
- linked `CurSearch` to `IncSearch,
- tried to optimize `StatusLine` and `StatuslineNC` in 16c and 8c,
- merged most 8c groups with the rest to avoid duplicates,
- reorganized the template in hopefully meaningful foldable sections.

Are there any other problems left (or created)?

habamax commented:
@romainl quickly checked it -- I like it (even more in 256c with black bg).

Will do more tomorrow.

neutaaaaan commented:
There is something weird going on with `t_Co=8`.
![image](https://user-images.githubusercontent.com/35198994/219863980-33af7fc2-78ca-4500-b9ea-f2641cb4a88b.png)

romainl commented:
@neutaaaaan could you be more specific?

neutaaaaan commented:
Probably an issue with the template itself: vim has a tendency to automagically make bold things use the brighter half of the palette despite t_Co being set to 8, and in this case `Normal` gets stuck using ansi15 and bold despite `cterm` being empty (you can even reset it manually, it won't budge) and `ctermfg` being set to 7, as it should be. 
Manually setting `ctermfg` to 7 fixes the bad behaviour.

romainl commented:
Oh, the bolded white? Yeah, I noticed it too and toyed with the same solution. Wait for a new commit :-).

habamax commented:
gui, 256c and 16c -- so far so good.

romainl commented:
So I made a few tweaks, mostly to get 8c to a workable state and improve cohesion.

I consider this done on my end.

Tell me if there are still problems.

neutaaaaan commented:
![image](https://user-images.githubusercontent.com/35198994/219963986-23480727-0bab-49af-9c26-42d37709e58e.png)
`Cursorline` somehow leaks through during diffs.

habamax commented:
@neutaaaaan do you mean CursorLineNr or CursorLineFold or CursorLineSign? 

t_Co=8:
![image](https://user-images.githubusercontent.com/234774/219976405-c4d05d06-0ce8-40b7-aceb-87f6d81d5b38.png)

t_Co=16
![image](https://user-images.githubusercontent.com/234774/219976415-3ae12244-fbb6-441f-9e1f-0783a58dac0d.png)

t_Co=256
![image](https://user-images.githubusercontent.com/234774/219976445-9c41d388-d9db-44fa-85f2-c31f9d2e0185.png)

neutaaaaan commented:
`CursorLineFold`

habamax commented:
I guess this is "by design" of zaibatsu. I personally make cusrolinefold/nr/sign same bg as normal but this is up to the author.

romainl commented:
Hmm, yeah, I linked all `CursorLine*` to `CursorLine`. As I understand it, the purpose of those highlight groups is precisely to extend the `CursorLine` into those columns so that the curent line is always easy to spot.

As for `Diff*`, I still don't like their interplay with `CursorLine` and I am quite tempted to use my own accessible diff scheme instead. Thoughts?

habamax commented:
Looks ok to me in screenshots.

romainl commented:
@habamax I just pushed the new diff colors if you want to try them.

habamax commented:
Search clashes with DiffText in GUI/256 (searching for `ord`):

Before:
![image](https://user-images.githubusercontent.com/234774/220222916-6cfc99c0-a987-48ab-ade6-ba9deae01122.png)

While searching:
![image](https://user-images.githubusercontent.com/234774/220222819-c1e163ca-e818-4789-bb15-1a4266b92e19.png)

Ok for 16/8 colors:
![image](https://user-images.githubusercontent.com/234774/220223015-3ffde21e-d3bd-4854-b7d7-5e965797a90f.png)

romainl commented:
@habamax not sure why I used cyan for GUI instead of the neutral1. Fixed.

habamax commented:
@romainl looks good to me, ship it :)

romainl commented:
Holala… the excitement…

romainl commented:
Here we go.

gdupras commented:
I really like this new colorscheme! One issue I've found so far however is that NonText looks like Normal.

romainl commented:
@gdupras I see what you mean. Using one of the neutralN would work better. Or do you have another suggestion?

gdupras commented:
I didn't really have any color in mind but after trying out neutral2 and neutral4 as you suggested I think they would work fine. neutral1 is too close to Normal and neutral5 is too close to the background color IMO. neutral3 is already used for SpecialKey and Ignore.

romainl commented:
Neutral1 and neutral3 are my most portable "grays" so I think neutral3 will do.

256c:
<img width="638" alt="Capture d’écran 2023-02-22 à 07 42 19" src="https://user-images.githubusercontent.com/344335/220543201-0a68ebca-0f0c-4b07-9f44-742fe1c574b4.png">

GUI:
<img width="911" alt="Capture d’écran 2023-02-22 à 07 43 53" src="https://user-images.githubusercontent.com/344335/220543469-44ea90ac-7323-4762-a8c3-bb8357d2960b.png">
